### DISCUSSION TITLE
Can I add a custom command automatically after a command using the lua script?

### DISCUSSION THREAD

**Discussion by xxnuo** [chrisant996/clink#328] [Q&A] [answered]
**I want to pipe stdout and stderr into a program, the following command works.**

cmd(I am using git): 

```bash
ls 2>&1 | less
```

output: 

```bash
bin
config
git
registry
(END)
```

**Can i and How do i write a lua script to automatically add the following commands ` 2>&1 | less` and get the same effect just using `ls`?**

**Comment by pukkandan** [comment]
`onfilterinput` can be used to replace the input line at execution time. Eg:

```lua
clink.onfilterinput(function (line)
    if line == 'ls' then
        return 'ls 2>&1 | less', false
    end
end)
```

**Reply by xxnuo** [reply]
Thanks! I will have a try.
I would appreciate it if you had any solutions to match all the exes and commands(including `ls` `git` `ping` and any others).

**Reply by xxnuo** [reply]
Thanks~
I found `clink.onfilterinput` can satisfy my needs.

**Comment by chrisant996** [comment, marked as answer]
Clink isn't needed for the stated situation.  You could use `doskey` to define an alias.

`doskey ls=ls $* 2^>^&1 ^| less`

I forget offhand whether `^` will successfully escape the redirection symbols when defining an alias.

There are many ways to do this sort of thing with doskey aliases.

Doskey is part of the Windows OS.  Looking up documentation and examples for doskey should help.

**Reply by xxnuo** [reply]
Thanks for answering!
It looks like doskey can make alias for specific commands with the pipe using `$B` as `|`.
I am looking for a universal solution for all the exes and commands (including `ls` `git` `ping` and so on) , I will read the doskey's documentation to see if it supports wildcard.
I would appreciate it if you had any solutions to match them.

**Reply by xxnuo** [reply]
Thanks~
I found `clink.onfilterinput` can satisfy my needs!

**Reply by chrisant996** [reply]
> Thanks for answering!
> It looks like doskey can make alias for specific commands with the pipe using `$B` as `|`.
> I am looking for a universal solution for all the exes and commands (including `ls` `git` `ping` and so on) , I will read the doskey's documentation to see if it supports wildcard.
> I would appreciate it if you had any solutions to match them.

What do you mean by "if it supports wildcard"?  The `$*` inserts any arguments that are typed.

`clink.onfilterinput()` can also work.

Some advantages of doskey aliases are:
- they can work in cmd even without Clink
- an input file can define many aliases
- works in compound commands, e.g. `program1 & ls & program3`

Some disadvantages of onfilterinput are:
- more complicated
- requires custom parsing code
- may be difficult to make it work properly in compound commands, e.g. `program1 & ls & program3`
- mistakes can potentially make it impossible to enter some commands

**Reply by xxnuo** [reply]
Sorry, I did not express it clearly. 

My purpose is to write a program using Rust to capture the output (stdout & stderr) of all commands and exes (that's why i try to use wildcard to match but i didn't),

and then output the outputs(stdout & stderr) after processing(such as translate english into chinese):

Inappropriate eg. Replace word `ok` to `fine`:

> `cargo` or `cargo.exe`: the package manager of rust language

```bash
> cargo test
running 1 test
test tests::one_result ... ok
test result: ok. 1 passed;
```

to

```bash
> cargo test
running 1 test
test tests::one_result ... fine
test result: fine. 1 passed;

```

There are two optional solutions:

1. add precommand (eg.sudo, my program execute the commands and output)

`myprogram.exe cargo test`

2. pipe output to my program (my program just handles the outputs)

`cargo test | myprogram.exe`

I want to automatically complete it that to add precommand or pipe command.

Now I use method 1 with clink using `precommand.lua`:

```lua
clink.onfilterinput(function (line)
	return "myprogram "..line,false
end)
```

My purpose looks like adding a custom shell function,
So I refer to some shell implementation.
But to be honest, I am not good at English and programming languages, except Python and JS.
I have been using clink for a while and it works very well!
So I want to achieve my purpose through the script function of clink instead of modifying a shell program or doskey(I don't know whether I can use something like ` doskey *= `  or ` doskey $*= `? to handle all the commands before).

Sorry to say so much.
All in all, Thank you for informing the shortcomings of `clink.onfilterinput()`,
so far the scripts above can basically work correctly.

And I will try to use doskey.

I'm not good at Bash programming, and I am very grateful to get suggestions from skilled programmers like you.

For the first time to discuss on GitHub, thank you very much for your answer.

**Reply by chrisant996** [reply]
Oh, the broader context helped to clarify it, thank you.

I agree that to add ` 2>&1 | less` to all commands, using `clink.onfilterinput()` would be needed.  I wonder whether applying it to "all" may cause problems, though.

The things to be aware of:
- Mistakes in the parsing code can potentially cause unexpected malfunctions, or cause commands to behave very differently.
- Should it really apply to "all" commands?  What if someone runs `foo | bar`?  Then it becomes `foo | bar 2>&1 | less`, which is probably not what's desired (for example, the `2>&1` applies to `bar` rather than to `foo`).  Or what if someone runs `guiprogram`, then it becomes `guiprogram 2>&1 | less`, which may be nonsensical for many GUI programs, and may cause some GUI programs to behave differently than expected.
- Multiple commands separated by `&` won't behave as expected.  Parsing them is more complicated than it appears.  Clink really needs to do the parsing and pass an additional input to the callback function, providing a table of parsed commands in the input line, similar to [here](https://chrisant996.github.io/clink/clink.html#setting-a-classifier-function-for-the-whole-input-line).

Doskey macros are a more predictable and safe approach.  But depending on the specific goals and acceptable set of side effects, `clink.onfilterinput()` may work as well.
